integr8ly/integreatly-operator
# INTLY-3686 Perform backup of products before upgrade

# Description
Implement components to perform and wait for backups on AWS resources or by creating
`Jobs` of backup `CronJobs`.
Add logic to perform the backups required for the different products before approving
an `InstallPlan`

## Type of change


## Steps to verify

The behaviour of this feature is different for installations using cluster storage than AWS storage. In order to fully verify this PR, it must be done on both kinds of installations. Repeat the following steps with: 1. `useClusterStorage = true` 2. `useClusterStorage = false`

* In a cluster, install RHMI and verify that the installation is not affected (no backup jobs are created either from CronJobs or PostgresSnapshot/RedisSnapshot)
* For each product listed in the [backup SOP](https://github.com/RHCloudServices/integreatly-help/tree/master/sops/2.x/backup):
  * Simulate a pending upgrade by setting the `approved` flag in the product `InstallPlan` to `false`
  * The operator should trigger the backups for the product (**Note: for installations where `useClusterStorage == true` no AWS backup should be performed**). Verify that:
    i. For AWS Postgres backups -> A `PostgresSnapshot` CR is created in the `redhat-rhmi-operator` namespace
    ii. For AWS Redis backups -> A `RedisSnapshot` CR is created in the `redhat-rhmi-operator` namespace
    iii. For CronJob based backups -> A job is created in the product namespace based on the backup CronJob
  * Verify that the backups successfully finish and the `InstallPlan` is approved
* For each product **not** listed in the backup SOP:
  * Simulate a pending upgrade as done in the previous steps
  * Verify that no backup job is triggered and the upgrade is approved

## Checklist
